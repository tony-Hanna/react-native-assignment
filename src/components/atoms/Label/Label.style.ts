import { StyleSheet } from "react-native";
export const labelStyles = StyleSheet.create({
    label: {
        fontSize: 13,
        marginBottom: 4,
    },
    errorLabel: {
        color: "red",
    },
})