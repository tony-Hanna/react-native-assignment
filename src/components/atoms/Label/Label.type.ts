import { StyleProp, TextStyle } from "react-native"
export interface LabelProps {
    label: string
    style?: StyleProp<TextStyle>
}
