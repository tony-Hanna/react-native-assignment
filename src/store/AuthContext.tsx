import React, { createContext, useContext, useState } from 'react';

const AuthContext = createContext<{
  isAuthenticated: boolean;
  setIsAuthenticated: (auth: boolean) => void;
}>({
  isAuthenticated: false,
  setIsAuthenticated: () => {},
});

export const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  const [isAuthenticated, setIsAuthenticated] = useState(false);

  return (
    <AuthContext.Provider value={{ isAuthenticated, setIsAuthenticated }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => useContext(AuthContext);
