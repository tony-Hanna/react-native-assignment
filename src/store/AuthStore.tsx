import { create } from 'zustand';
import RNSecureStorage, {ACCESSIBLE} from 'rn-secure-storage';
import { createJSONStorage, persist, StateStorage } from 'zustand/middleware';

const SecureStorage: StateStorage = {
  setItem: async (name, value) => {
    try {
      await RNSecureStorage.setItem(name, value, {
        accessible: ACCESSIBLE.WHEN_UNLOCKED,
      });
    } catch (error) {
      console.error('Error setting secure storage item:', error);
    }
  },

  getItem: async name => {
    try {
      const value = await RNSecureStorage.getItem(name);
      return value ?? null;
    } catch (error: any) {
      if (error?.message?.includes(`Value for ${name} does not exist`)) {
         console.log(`Secure storage item '${name}' not found on startup.`);
         return null;
      } 
      console.error('Error getting secure storage item:', error);
      return null;
    }
  },

  removeItem: async name => {
    try {
      await RNSecureStorage.removeItem(name);
    } catch (error) {
      console.error('Error removing secure storage item:', error);
    }
  },
};

type AuthState = {
  userId: string | null;
  accessToken: string | null;
  refreshToken: string | null;
  setTokens: (accessToken: string, refreshToken: string) => void;
  clearTokens: () => void;
  hasStoreLoaded: boolean;
  setHasStoreLoaded: () => void;
  setUserId: (userId: string) => void;
}

export const useAuthStore = create<AuthState>()(
  persist(
    (set) => ({
      accessToken: null,
      refreshToken: null,
      userId: null,
      hasStoreLoaded: false,
      setHasStoreLoaded: () => set({ hasStoreLoaded: true }),
      setTokens: (accessToken, refreshToken) =>
        set({ accessToken, refreshToken }),
      clearTokens: () => set({ accessToken: null, refreshToken: null }),
      setUserId: (userId: string) => set({ userId }),
    }),
    {
      name: 'auth-storage',
      onRehydrateStorage: () => (state) => {
        // Ensure we always set hasStoreLoaded to true, even if rehydration fails
        if (state) {
          state.setHasStoreLoaded();
        } else {
          // If state is null, something went wrong with rehydration
          console.warn('Auth store rehydration failed');
          useAuthStore.getState().setHasStoreLoaded();
        }
      },
      storage: createJSONStorage(() => SecureStorage),
    },
  ),
);
