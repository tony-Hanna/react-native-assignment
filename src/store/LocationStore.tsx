import { create } from 'zustand';
import { Region } from 'react-native-maps';

type LocationStore = {
  location: Region | null;
  setLocation: (location: Region) => void;
  clearLocation: () => void;
};

export const useLocationStore = create<LocationStore>((set) => ({
  location: null,
  setLocation: (location) => set({ location }),
  clearLocation: () => set({ location: null }),
}));
