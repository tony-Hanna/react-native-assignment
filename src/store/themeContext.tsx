import { useState, useContext, createContext, useMemo, useCallback, memo } from "react";

export interface Theme {
  background: string;
  text: string;
  gradient: string[]
}

interface ThemeContextType {
  theme: Theme;
  isDark: boolean;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextType>({} as ThemeContextType);

const lightTheme: Theme = {
  background: "#ffffff",
  text: "#000000",
  gradient : ['#fde6d5', '#dfd4ff']
};

const darkTheme: Theme = {
  background: "#000000",
  text: "#ffffff",
  gradient : ['#0f2027', '#243B55']
};

export const ThemeProvider = ({ children }: { children: React.ReactNode }) => {
  const [isDark, setIsDark] = useState(false);

  const toggleTheme = () => setIsDark(prev => !prev);


  const theme = isDark ? darkTheme : lightTheme;

  const contextValue = useMemo(() => ({
    theme,
    isDark,
    toggleTheme
  }), [theme, isDark, toggleTheme]);

  return (
    <ThemeContext.Provider value={contextValue}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useTheme = () => useContext(ThemeContext);
  