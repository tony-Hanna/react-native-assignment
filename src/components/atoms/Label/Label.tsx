import { Text } from "react-native"
import { LabelProps } from "./Label.type"
import { labelStyles } from "./Label.style"
const Label : React.FC<LabelProps> = ({label, style}) => {
    return (
        <Text style={[style ,labelStyles.label]}>
            {label}
        </Text>
    )
}

export {Label}
