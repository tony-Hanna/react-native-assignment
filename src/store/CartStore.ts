import { create } from "zustand";
import { MMKV } from 'react-native-mmkv';
import { persist, createJSONStorage } from 'zustand/middleware';
import { CartItem } from "../screens/Cart/Cart.type";
import { useAuthStore } from "./AuthStore";
// Initialize MMKV storage
const storage = new MMKV();

// Helper to get the current user-specific storage key
const getCartStorageKey = () => {
  // Use getState to avoid hook-in-hook issues
  const userId = require('./AuthStore').useAuthStore.getState().userId;
  return userId ? `cart-storage-${userId}` : 'cart-storage-guest';
};

// Define the cart store state and actions
interface CartStore {
  items: CartItem[];
  addItem: (item: Omit<CartItem, 'quantity'>) => void;
  removeItem: (itemId: string) => void;
  updateQuantity: (itemId: string, quantity: number) => void;
  clearCart: () => void;
  getTotalItems: () => number;
  getTotalPrice: () => number;
  isInCart: (itemId: string) => boolean;
}

// Create the cart store with user-specific persistence
export const useCartStore = create<CartStore>()(
  persist(
    (set, get) => ({
      items: [],

      addItem: (item) => {
        set((state) => {
          const existingItem = state.items.find((i) => i._id === item._id);
          
          if (existingItem) {
            // If item exists, increase quantity
            return {
              items: state.items.map((i) =>
                i._id === item._id
                  ? { ...i, quantity: i.quantity + 1 }
                  : i
              ),
            };
          }
          
          // If item doesn't exist, add new item with quantity 1
          return {
            items: [...state.items, { ...item, quantity: 1 }],
          };
        });
      },

      removeItem: (itemId) => {
        set((state) => ({
          items: state.items.filter((item) => item._id !== itemId),
        }));
      },

      updateQuantity: (itemId, quantity) => {
        if (quantity < 1) {
          get().removeItem(itemId);
          return;
        }

        set((state) => ({
          items: state.items.map((item) =>
            item._id === itemId ? { ...item, quantity } : item
          ),
        }));
      },

      clearCart: () => {
        set({ items: [] });
      },

      getTotalItems: () => {
        return get().items.reduce((total, item) => total + item.quantity, 0);
      },

      getTotalPrice: () => {
        return get().items.reduce(
          (total, item) => total + item.price * item.quantity,
          0
        );
      },

      isInCart: (itemId) => {
        return get().items.some((item) => item._id === itemId);
      },
    }),
    {
      name: getCartStorageKey(),
      storage: createJSONStorage(() => ({
        getItem: (name) => {
          const value = storage.getString(name);
          return value ? JSON.parse(value) : null;
        },
        setItem: (name, value) => {
          storage.set(name, JSON.stringify(value));
        },
        removeItem: (name) => {
          storage.delete(name);
        },
      })),
    }
  )
);