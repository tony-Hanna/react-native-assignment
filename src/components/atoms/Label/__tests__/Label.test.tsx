import React from 'react';
import { Label } from '../Label';
import { render } from "@testing-library/react-native";
import { ThemeProvider } from '../../../../store/themeContext';
const props = {
    label: 'some text'
}
it("renders CustomText correctly", () => {
    const { getByText, toJSON } = render(
        <ThemeProvider>
            <Label {...props}/>
        </ThemeProvider>
    );
    expect(getByText('some text')).toBeTruthy()
    expect(toJSON()).toMatchSnapshot()
});
