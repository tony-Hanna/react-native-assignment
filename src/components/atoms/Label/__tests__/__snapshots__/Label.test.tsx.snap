// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CustomText correctly 1`] = `
<Text
  style={
    [
      undefined,
      {
        "color": "#000000",
        "fontFamily": "Roboto-Medium",
        "fontSize": 13,
        "marginBottom": 4,
      },
    ]
  }
>
  some text
</Text>
`;
